### ISSUE TITLE
Freeze on FirebaseRemoteConfig.ActivateFetched() Unity 2017.4.6f1.

### ISSUE THREAD

**Issue by DenizPiri** [firebase/quickstart-unity#182]
When calling FirebaseRemoteConfig.ActivateFetched() at the start of the app, at first frame, the app gets stuck in launch screen image. There seem to be no apparent errors in "adb logcat" output. This normally happens with 95% chance.

Deferring the ActivateFetched call by 100ms reduces the chances of getting this launch freeze a lot. In this case, chances of freezing is around 5%.

This is how my current initialization code looks like:

```
Firebase.FirebaseApp.CheckAndFixDependenciesAsync().ContinueWith(task =>
        {
            var dependencyStatus = task.Result;
            MainThreadDeferer.Instance.CallOnUnityThread(() =>
            {
                if (Debug.isDebugBuild)
                    Debug.LogFormat("Firebase.DependencyStatus: {0}", dependencyStatus);

                if (dependencyStatus == Firebase.DependencyStatus.Available)
                {
                    // Fetch and activate firebase remote config data.
                    Run.AfterRealtime(0.1f, () => // WARNING: Without this ActivateFetched gets stuck on Android sometimes.
                    {
                        Firebase.RemoteConfig.FirebaseRemoteConfig.ActivateFetched();
                        Firebase.RemoteConfig.FirebaseRemoteConfig.FetchAsync().ContinueWith(task2 =>
                        {
                            MainThreadDeferer.Instance.CallOnUnityThread(() =>
                            {
                                if (Debug.isDebugBuild)
                                    Debug.LogFormat("Fetched firebase remote config");
                                Firebase.RemoteConfig.FirebaseRemoteConfig.ActivateFetched();
                            });
                        });
                    });
                }
                else
                {
                    UnityEngine.Debug.LogError(System.String.Format(
                      "Could not resolve all Firebase dependencies: {0}", dependencyStatus));
                    // Firebase Unity SDK is not safe to use here.
                }
            });
        });
```

MainThreadDeferer.Instance.CallOnUnityThread calls are there to defer the callbacks to main unity thread, just to make sure that task callbacks are not called in a separate thread.

Runtime I am on is Mono with .NET 2.0 Subset. Stripping level is set to "micro mscorlib". Min API level is Android 4.1, the target is Android 8.0. Building for ARMv7 arch.

I am puzzled by what might be causing the problem of being stuck. I wish firebase unity integration code was open source, would be pretty easy to understand what is going on. I assume that call is using some async API internally and waiting for the result in a while loop, just to be able to return the "bool" value. Then again, the Java API for activateFetched is not an async call. However, maybe, to get the call to the right thread, some async operation is going on there.

Would be great to get some help with this. I spent half the day battling with the error, still no clear solution. Deferring the ActivateFetched call by a constant time reduces chances of the problem happening, but doesn't fully solve it.

**Comment by a-maurice** [comment]
Thanks for alerting us to this issue.  I'm not sure why it is hanging in that case, that's definitely odd behavior that we'll need to sort out.

That said, you should not need to do that first ActivateFetched() that your sample code has, as it is before the FetchAsync().  The basic flow it typically FetchAsync() -> after some time, the result are ready, but it still uses default / older values -> ActivateFetched() -> it now uses up to date values.  So doing the ActivateFetched() prior to a Fetch should effectively be a no op, but it seems like something is going wrong there, which we will want to fix.

Also, both FetchAsync() and ActivateFetched() should be safe to call from any thread, not necessarily just the Unity main thread.  In general, the only Firebase call that needs to be done from the Unity main thread is CheckAndFixDependenciesAsync, but if you need to do any interaction with game objects or other Unity calls, you will want to do something similar to what you have, with a dispatcher.

And unrelated, you will likely want to use FirebaseRemoteConfig.SetDefaults() to set default values for your remote config fields, prior to calling FetchAsync().  These are useful for when the user doesn't have internet connection, or if you just don't have anything set in the console for that value.

So, if you could try dropping that ActivateFetched() prior to the FetchAsync(), and see if that helps.  If you are still having issues after that, please let us know!

**Comment by DenizPiri** [comment]
I built a fresh project by only importing FirebaseAnalytics.unitypackage and FirebaseRemoteConfig.unitypackage.
I used the same initialization code from our game. Tried to start the app about 30 times, the freeze doesn't occur.
However, when I move the FirebaseRemoteConfig calls outside of CheckAndFixDependenciesAsync callback. Same freeze occurs with a very high chance. This is the current code and last thing I see on logcat is "Firebase init 2". After that point, there are no errors and no other logs by the app.

```
Debug.LogFormat("Firebase init 0");
        Firebase.FirebaseApp.CheckAndFixDependenciesAsync().ContinueWith(task =>
        {
            var dependencyStatus = task.Result;
            if (dependencyStatus == Firebase.DependencyStatus.Available)
            {
                
            }
            else
            {
                UnityEngine.Debug.LogError(System.String.Format(
                  "Could not resolve all Firebase dependencies: {0}", dependencyStatus));
            }
        });
        Debug.LogFormat("Firebase init 1");

        Debug.LogFormat("Firebase init 2");
        Firebase.RemoteConfig.FirebaseRemoteConfig.SetDefaults(new Dictionary<string, object>());
        Debug.LogFormat("Firebase init 3");
        Firebase.RemoteConfig.FirebaseRemoteConfig.FetchAsync().ContinueWith(task2 =>
        {
            Debug.LogFormat("Firebase init 3.1");
            Firebase.RemoteConfig.FirebaseRemoteConfig.ActivateFetched();
            Debug.LogFormat("Firebase init 3.2");
        });
        Debug.LogFormat("Firebase init 4");
```

I believe whatever is causing the freeze with a very high chance, in this case, is causing the freeze in our app as well. I assume remote config tries to wait on a mutex or an event that never gets triggered. Or maybe some kind of deadlock because of locking order?

Why isn't the firebase unity adapter opensource? Isn't it just a bridge to the Android and iOS clients anyway? Would be great to see it's code, to get to know what I am dealing with. Currently, all I can do is to change call ordering, timing etc to hope that this freeze will not occur.

Here is the sample project: https://drive.google.com/open?id=1bTn811DcqAWGNgYWfw9hEMBUq8pUzR2c

**Comment by DenizPiri** [comment]
I think I have found a dirty workaround. I create a new thread to call FetchAsync. Thinking that if it locks up, it won't lock up the unity thread. Also if one thread locks up, a second one is started 5 secs later, then 3rd attempt 15 secs later.

However, this is a super scary and dirty solution in my opinion. It is highly possible that game might do an early call to FirebaseRemoteConfig.GetValue and same lockup might occur. Without being able to look into the sources of firebase unity SDK, I can't predict the behavior.

**Comment by DenizPiri** [comment]
I found this issue here, that seems to be related to the issue I am having: https://stackoverflow.com/questions/41959490/firebase-makes-tasks-execute-only-on-main-thread-causing-unity-to-freeze

**Comment by DenizPiri** [comment]
Ok, I have been testing things with Development builds. Now that I switched to release builds, in our game even calling Firebase.FirebaseApp.CheckAndFixDependenciesAsync() causes a lock up. In this case not inside the call, but somewhere else (don't know where, yet). Possibly inside Unity.Tasks library?

I started to decompile Unity.Tasks library, to understand what is going on. I ran into a possible freeze. in Task.Wait() call. It is locking the mutex with Monitor.Enter() and right after that it checks IsCompleted property, IsCompleted implementation tries to lock the exact same mutex again. According to my knowledge, the default mutexes are not reentrant, however not sure how Unity's Mono C# runtime handles it.

**Comment by a-maurice** [comment]
Ah, you should not be using any Firebase calls until after CheckAndFixDependenciesAsync has finished.  What it does is resolve any Android dependencies, which could potentially involve prompting the user to the Play store to update their Google Play Services, and trying to access Firebase features before that has finished can lead to unexpected behavior.

What we recommend, and what the testapps here do is:
Firebase.FirebaseApp.CheckAndFixDependenciesAsync().ContinueWith(task => {
  var dependencyStatus = task.Result;
  if (dependencyStatus == Firebase.DependencyStatus.Available) {
    // Initialize Firebase here
  } else {
    UnityEngine.Debug.LogError(System.String.Format(
                  "Could not resolve all Firebase dependencies: {0}", dependencyStatus));
  }
});

The reason for the odd behavior is that CheckAndFixDependenciesAsync potentially creates, and then destroys the default FirebaseApp, which Remote Config also uses.  What is likely happening in this case is that the FirebaseApp that Remote Config is trying to use is getting destroyed as CheckAndFixDependencies is finishing, which is causing the freeze.

It sounds like moving that initialization logic into the continuation of CheckAndFixDependenciesAsync worked for you before, but if it does not, please let me know.

**Comment by DenizPiri** [comment]
Ok, I feel like I am not getting closer to figure it out using disassembled code.

I read the Tasks library by Parse, found it in an older version of their unity.net library. It overall seems to be fine. Found out that mutexes in C# are recursive by default. I see that there is a class called UnitySynchronizationContext in Firebase.Platform. I assumed that it should be used by default, however, I can't see any reference to it anywhere else. It might also be because of the decompiler.

One thing I noticed though is that, as far as I can tell from the decompiled code, SynchronizationContextBehavoir (typo in code) runs callbacks in Unity thread not in Update() callback but starts a coroutine at it's Start, and the coroutine invokes the callbacks. However, as far as I know, coroutines on DontDestroyOnLoad objects stop between scene changes. Maybe that behavior changed in newer versions, but as far as I know, this was the case in older unity versions. As for the couroutine itself, dotpeek fails to decompile that, so can't tell what is going on there.

Overall, I expect freezes to be originating from tasks library and **SynchronizationContextBehavoir**. I wish this lib didn't depend on .net Tasks at all. I don't know what more I can do at this point. It truly sucks that firebase lib is not open source. My current approach will be to remove firebase from our Unity based projects, as we cannot afford to release a game that fully freezes at any point. In meantime, I will eagerly wait for a fix by the Google Firebase team ⌛️⌛️⌛️

Cheers!

**Comment by DenizPiri** [comment]
@a-maurice As I pointed out, I get freezes even by simply using CheckAndFixDependenciesAsync(), especially that combined with a LoadScene that comes right after it. If I defer the LoadScene call by a second or two, then it works fine.

However, I really need to initialize firebase, start fetching the remote data and start loading the main menu all at once without even skipping a frame. Therefore, I would prefer it Firebase would use a normal TaskScheduler and I would be getting the init callback while unity is loading the main menu, and as a result, remote config fetching would start while unity is busy loading.

**Comment by a-maurice** [comment]
I've tried a handful of things, including the sample project you provided, and in release builds I have still been unable to reproduce a freeze with just CheckAndFixDependenciesAsync(), including with a call to LoadScene afterwards.  If I do the FetchAsync() simultaneously, I do see a freeze, but as I mentioned that behavior is undefined, as the intention is that the task returned by CheckAndFixDependenciesAsync() needs to be finished before using any Firebase features, as otherwise some functionality might be missing, causing other failures.  Have you been able to reproduce this in the remote config testapp that we provide?

Regarding your point above about SynchronizationContextBehavoir, that does seem like it might be an issue.  There is actually a different MonoBehaviour that we ordinarily use for running functions on the main thread, which does use Update, but as you pointed out that SynchronizationContextBehavoir does not.  We'll look into fixing this, but unfortunately I'm not sure if that will resolve the issue that you are seeing.

And thanks for the interest in open sourcing the Unity SDK.  As you may have seen, both the iOS and Android SDKs have been moving more of their code to open source.  The Unity libraries rely heavily on code generation, which makes it a bit harder than just exposing the cs files, but it's great to know that there is interest in it.

**Comment by DenizPiri** [comment]
In my experience, it is never ok for an API to lock up the thread. If it does, it means that there is a problem with the use of synchronization primitives.
In the sample project I provided, I placed the functions in that order just to increase the chances of the freeze. We don't do the initialization that way in our app.

I didn't try the remote config testapp you mention. Now I will try to have a reproducible freeze, by removing things from our app, instead of creating a new sample. Will post here when I have something.

**Comment by DenizPiri** [comment]
Ok, after spending 2 more days on it, I distilled our big project into a small case where I can create the freeze. It basically occurs when Firebase.FirebaseApp.CheckAndFixDependenciesAsync() called followed by a SceneManager.LoadScene() call, and then in the loaded scene, calling FirebaseRemoteConfig.GetValue(), it instantly freezes at that point. Once again, no exceptions, no logs.

Here is a sample project showing the freeze: https://drive.google.com/open?id=1-0N9M85SeUF4nNi1JJ716vhUOziyuMmn

If I move the SceneManager.LoadScene call within the ContinueWith callback of Firebase.FirebaseApp.CheckAndFixDependenciesAsync(), the freeze doesn't occur.

As I understand, only proper solution from our side is to make an independent load scene for firebase, and initialize it before we initialize anything else, but since it is async, it will go from our launch screen image to black screen (firebase initialize scene) then the actual intro scene, where we start initializing things which may be using firebase.
Or wrap the whole firebase remote config API, cache the previous results ourselves, and use the previous value results if firebase callback is not called yet. Definitely not going for this option...

We are probably not going with a workaround like this since my trust in Firebase Unity SDK got fully crushed. We will opt for not using Firebase for our Unity based projects until all these possible cases of thread lock freezes disappear. As I mentioned before, under no condition an API should just block. Especially, when it is not a threading primitive itself or an async object. Also nowhere in Firebase Unity SDK docs it is mentioned that calling Firebase functions before CheckAndFixDependenciesAsync callback returns might lead to thread freezes.

Another suggestion would be to rename the CheckAndFixDependenciesAsync function. That naming completely sounds like an operation that is just there to check the dependencies. Not something that initializes an SDK and SDK being usable only after that async op completes.

**Comment by a-maurice** [comment]
Ah, that definitely clarifies the issue that you are seeing, as the loaded scene is trying to use Firebase as CheckAndFixDependenciesAsync is still running, so thank you for including that.  And yes, if you want to start using Remote Config as soon as possible, then you would need to wait for CheckAndFixDependenciesAsync() to finish, as otherwise it is possible that the Android device does not have an up to date Google Play Services, which could cause exceptions being thrown, and other issues.

There are however alternatives to waiting for the initialization, if that is a concern.  For example, you could use some default values for whatever fields you need, and then when both CheckAndFixDependenciesAsync and FetchAsync are finished, replace those with the fetched values.  However, if you were looking to get the values before starting anything, then you would need to wait for both CheckAndFixDependenciesAsync and FetchAsync to be finished.

We do try to call out the need for CheckAndFixDependenciesAsync(), for example https://firebase.google.com/docs/remote-config/use-config-unity#initialize_the_sdk, but clearly we need to be doing a better job, so thank you for bringing that to our attention.  And also thank you for bringing to our attention this freezing issue.  While it is not the intention that other calls happen before CheckAndFixDepenciesAsync has finished, it should be throwing an exception, or something that can be handled, instead of potentially locking the game.  We'll definitely work on getting a fix for that issue out.

**Comment by DenizPiri** [comment]
Ok, then the API is just wrong. Definitely not made for the way Unity works and the way most Unity projects are structured.

Most unity projects, in their first scene, have lots of objects that start doing things in their Awake/Start functions. Some of these things will naturally depend on a remote config value or trigger an analytics event. In Unity, there is no Init() function that runs before your scene starts or while the engine starts up. Best bet is to use a transient scene, but then if you spend more than a frame in that scene before your next LoadScene function, you end up clearing the launch image and screen goes black. So using CheckAndFixDependenciesAsync() in a transient scene and using that as the first scene which loads the actual intro/menu scene when Firebase initialization is done is also not an option, loading sequence ends up looking really ugly.

My current solution is to use Firebase.FirebaseApp.CheckDependencies in the first scene of the game, and do nothing before that. If it is Available, start a thread for fetching remote config (I also had a freeze with this previously, not trusting it anymore) proceed with using SceneManager.LoadScene. If CheckDependencies doesn't return a positive result, then call FixDependenciesAsync and in that case wait for its result to run the LoadScene call. However, as I understand even this is not a solution enough, because now I cannot know how the game will behave, for example, on an Amazon Kindle device. It might simply just freeze because of the uninitialized firebase. So I need to wrap all firebase APIs that I am using so that I can replace it with dummy implementations in case firebase didn't initialize properly.

I still can't declare this issue fully solved for our project's goal, we need more testing on more devices, especially the ones without Google Play Services. I also need to measure the timing of CheckDependencies function to make sure that we don't add an unnecessary 1-2 secs to our game's loading time.

From a API user's perspective I would expect the following:

- RemoteConfig APIs should work without having to wait for an async initialization process.
- RemoteConfig APIs should return previously activated results even when the firebase is not initialized
- RemoteConfig Fetch and all other Firebase calls should throw an exception in case they depend on firebase being initialized.
- Name of CheckAndFixDependenciesAsync should be changed. I have been programming for 20 years, for a function that must be called before touching rest of the API, it is one of the worst names I have seen. 😄 
- API should behave the same on all platforms. As in, if it throws an exception on Android, it should do the same on x64, even if it is in an emulated way. Same goes for iOS as well of course.
- If possible someone should review all threading code within firebase unity SDK, and make sure that it can't possibly lead to any lockups. Assuming the code is generated as you said, this should be relatively a small review task. I use Firebase on Android and iOS in our own engine, had no such troubles on those direct integrations. I would gladly do the review work if the lib was open source. 👍
- Quickstart samples should be fixed. Checking the remote_config sample, technically I can create the same freeze if I was quick enough to press the GUI buttons right after the app starts.

Please don't close this issue at least until the freeze issue is completely resolved.

Cheers!

**Comment by DenizPiri** [comment]
I measured the timing of CheckDependencies function, seems to be taking ~300ms on Galaxy S9, and 500 to **900ms** 🔥 on Xperia Z3. 😞😞😞

**Comment by stewartmiles** [comment]
@DenizPiri as @a-maurice pointed out, almost all Google, including Firebase, APIs on Android require [Google Play services](https://developers.google.com/android/guides/overview) at the moment.  This includes Remote Config which consists of the client library (bit you ship with your app) and the daemon / APK which is distributed to every Google Play Android device.  Each time the client SDK moves to the latest version the daemon / APK needs to be new enough to support the client.  We need to perform the check for the latest Google Play services / upgrade it to the required version to start using the Remote Config API on Android.

Now to each of your points...

- RemoteConfig APIs should work without having to wait for an async initialization process.

   - As described above, currently Google Play services is required for Firebase APIs.  You can't use the API until Google Play services is available as the client (in your application) can't talk to the server (running on device) until it has checked it is up to date and ready to be used.

- RemoteConfig APIs should return previously activated results even when the firebase is not initialized

   - As described above, Google Play services is required for Firebase APIs.  The client calls into the Google Play services APK on the device.  If initialization time was fast enough, I'm sure this could be made to operate as you expect, however that is currently not the case. 

- RemoteConfig Fetch and all other Firebase calls should throw an exception in case they depend on firebase being initialized.

   - This is indeed a bug.  All methods that are called prior to initialization being complete should be throwing exceptions.  We'll look into it.

- Name of CheckAndFixDependenciesAsync should be changed. I have been programming for 20 years, for a function that must be called before touching rest of the API, it is one of the worst names I have seen. 😄 

   - The function actually does what it says.  As described above, Google Play services is required to use Firebase on Android.  You can simply not call the method and start using the API (assuming the first API call is from the main thread), things should work until you run on a device with an out of date Google Play services and you'll be totally out of luck.

- API should behave the same on all platforms. As in, if it throws an exception on Android, it should do the same on x64, even if it is in an emulated way. Same goes for iOS as well of course.

   - Completely disagree here.  Different platforms are different beasts entirely.  On iOS and desktop we don't have a dependency upon Google Play services at all so forcing users to call it when they don't really have to seems a bit silly.  CheckAndFixDependenciesAsync works across platform (it's a stub on iOS & desktop) as a convenience, it's only required for Android.  Given the way the Firebase Android SDK is structured this is the best we can do at the moment.  Now things may change if check / update of Google Play services is improved (I've filed a bug against the owners) or we move parts of Firebase out of the Google Play services APK entirely at the expense of a larger Java client library size.

- If possible someone should review all threading code within firebase unity SDK, and make sure that it can't possibly lead to any lockups. Assuming the code is generated as you said, this should be relatively a small review task. I use Firebase on Android and iOS in our own engine, had no such troubles on those direct integrations. I would gladly do the review work if the lib was open source. 👍

   - We're actively working towards open sourcing our C++ and later Unity SDKs.  We look forward to your contributions.

- Quickstart samples should be fixed. Checking the remote_config sample, technically I can create the same freeze if I was quick enough to press the GUI buttons right after the app starts.

   - Thanks for the bug report, we'll take a look.

Now as you describe the process of performing a local RPC on device to check that Google Play services is up to date is very slow.   The 900ms seems crazy to me, running this in parallel seems like a long time as well especially if you have code dependent upon the library being initialized.  I've filed a bug against the team that owns this.

I recommend starting asynchronously initializing the SDK from whatever code paths you have that lead to a call into Firebase.  For example, while collaborating with @dzlier-gcp on [Firebase Unity Solutions](https://cloudplatform.googleblog.com/2018/06/new-gitHub-repo-using-firebase-to-add-cloud-based-features-to-games-built-on-unity.html) we came across a similar issue and solved it using this [FirebaseInitializer.cs](https://github.com/firebase/unity-solutions/blob/master/Assets/Firebase_Leaderboard/Scripts/FirebaseInitializer.cs) class which can be called from multiple entry points (i.e multiple behaviors) which then provides a signal when initialization is complete.

Given the constraints of the current Android SDK structure we would be happy to hear suggestions on how it could be improved.

**Comment by DenizPiri** [comment]
First of all, I would like to thank you both for taking your time and paying attention to the issues I am raising. This already put you way ahead of all the other open/closed source libs out there. 👏👏👏

Going to reply to the replies you have given.

- RemoteConfig APIs should work without having to wait for an async initialization process.
    - As described above, currently Google Play services is required for Firebase APIs. You can't use the API until Google Play services is available as the client (in your application) can't talk to the server (running on device) until it has checked it is up to date and ready to be used.
- RemoteConfig APIs should return previously activated results even when the firebase is not initialized
    - As described above, Google Play services is required for Firebase APIs. The client calls into the Google - Play services APK on the device. If initialization time was fast enough, I'm sure this could be made to operate as you expect, however that is currently not the case.
        - I understand, however I imagine that the remote config data is not stored by play services? In this case, what prevents firebase SDK from making remote config values accessible without having to initialize anything?
- RemoteConfig Fetch and all other Firebase calls should throw an exception in case they depend on firebase being initialized.

    - This is indeed a bug. All methods that are called prior to initialization being complete should be throwing exceptions. We'll look into it.
        - 👍 Great! I would still prefer it if RemoteConfig API's getValue function could work without any initialization. And without ever causing any freeze.
- Name of CheckAndFixDependenciesAsync should be changed. I have been programming for 20 years, for a function that must be called before touching rest of the API, it is one of the worst names I have seen. 😄

    - The function actually does what it says. As described above, Google Play services is required to use Firebase on Android. You can simply not call the method and start using the API (assuming the first API call is from the main thread), things should work until you run on a device with an out of date Google Play services and you'll be totally out of luck.
        - I don't agree here. I am assuming that with the first API call FirebaseApp would get created. Counting on this on other platforms simply sounds scary, if that blocks for 900ms or even 1ms, it is very bad having to block even 1ms because of reporting an analytic event. I would still strongly recommend renaming the function and making the API behave uniformly across all platforms.
- API should behave the same on all platforms. As in, if it throws an exception on Android, it should do the same on x64, even if it is in an emulated way. Same goes for iOS as well of course.

    - Completely disagree here. Different platforms are different beasts entirely. On iOS and desktop we don't have a dependency upon Google Play services at all so forcing users to call it when they don't really have to seems a bit silly. CheckAndFixDependenciesAsync works across platform (it's a stub on iOS & desktop) as a convenience, it's only required for Android. Given the way the Firebase Android SDK is structured this is the best we can do at the moment. Now things may change if check / update of Google Play services is improved (I've filed a bug against the owners) or we move parts of Firebase out of the Google Play services APK entirely at the expense of a larger Java client library size.
        - Ok, we are at a super big disagreement here. 😄 Would you like it if pthreads behaved differently on different across unix OSes? Or berkley sockets API behaving differently across platforms (ok, it sort of does 🤣🤣). APIs should be designed in ways that, if one function changes a state that another function depends on, this kind of behavior should never be platform specific. If it throws an exception or freezes in Android, it should do the same on iOS and Windows. Just so that, when your API's user doesn't need to deal with the different beasts you mentioned. If reporting analytic events would throw exception on Android before initializing, but would work on iOS, then it should either throw exception on all platforms, or wrappers should be written on platforms requiring initialization so that calls can be queued and executed later when initialization is done. And of course those wrappers should be on your side, not on the user's side.
Ideally, accessing dependent functions shouldn't even be allowed by API design. Such as Init() function returning an object which would let you access the rest of the API, etc, more of a hierarchical design. However, for the ease of use, I can understand not going that route. Basically, in most APIs that are not designed hierarchically, you have an internal state machine, as the user of the API, user needs to keep that state machine in mind while using it. Such as knowing that at certain points of code, API being uninitialized, initialized, connecting, destroyed, etc. Having a different state machine internally completely kills the point of making a unified API. Especially, when that state machine has some async components within.
Hopefuly I managed to explain my point of view a bit. I would much rather speak in person about code design. 😄
- If possible someone should review all threading code within firebase unity SDK, and make sure that it can't possibly lead to any lockups. Assuming the code is generated as you said, this should be relatively a small review task. I use Firebase on Android and iOS in our own engine, had no such troubles on those direct integrations. I would gladly do the review work if the lib was open source. 👍

    - We're actively working towards open sourcing our C++ and later Unity SDKs. We look forward to your contributions.
        - This would be great!
- Quickstart samples should be fixed. Checking the remote_config sample, technically I can create the same freeze if I was quick enough to press the GUI buttons right after the app starts.

    - Thanks for the bug report, we'll take a look.
        - 👍 


Anyway, once again thanks for your time. As I mentioned previoulsy, now using CheckDependencies as the first thing in the game and starting the actual game scene only after that works for us currently. Hopefully, we won't run into any other freeze issue. I will be looking forward for the freeze fix and a definitely a version where some parts of the API can work without requiring initialization. So, I will get back to my usual tasks, and will be on the lookout for future releases! ⌛️

Cheers! 🥂

**Comment by dzlier-gcp** [comment]
Thanks so much for the discussion on this topic! I am head-down on another task today so can't get too much into discussion at the moment, but wanted to mention that the RemoteConfig portion should definitely not have been pushed yet as it is not ready - it should only have been pushed to my personal fork while in development, so I am going to pull that back for now.

**Comment by DenizPiri** [comment]
@stewartmiles Great! Thanks!!!

While reading the changelog, I noticed this: "Firebase now throws an exception if any Firebase libraries are initialized while CheckAndFixDependenciesAsync() is still in progress.". This addresses my main complaint fully.

Thanks once again! ❤️

**Comment by stewartmiles** [comment]
Thanks @DenizPiri , @a-maurice sorted out the race with `CheckAndFixDependenciesAsync()`, I just get to announce stuff :+1:

**Comment by DenizPiri** [comment]
@a-maurice Thank you!!! ❤️

**Comment by AbhijitTaneja** [comment]
I am stilll facing this issue on Unity 5.6.3 p4 and firebase sdk 5.4.0. Need help!!!

**Comment by alexames** [comment]
@AbhijitTaneja, can you provide more detail? Did you confirm you're not calling any other Firebase API functions until `CheckAndFixDependenciesAsync()` has completed? What error are you getting?

**Comment by AbhijitTaneja** [comment]
Firebase.FirebaseApp.CheckAndFixDependenciesAsync().ContinueWith(task =>
        {
            var dependencyStatus = task.Result;
            if (dependencyStatus == Firebase.DependencyStatus.Available)
            {
                app = Firebase.FirebaseApp.DefaultInstance;

                Dictionary<string, object> defaultConfigDictionary = new Dictionary<string, object>();
                for (int i = 0; i < _ConfigKeyValuePair.Length; i++)
                    defaultConfigDictionary.Add(_ConfigKeyValuePair[i]._ConfigKey, "");

                Firebase.RemoteConfig.FirebaseRemoteConfig.SetDefaults(defaultConfigDictionary);
------------------------------**line**.................
            }
            else
                Debug.LogError(System.String.Format("Could not resolve all Firebase dependencies: {0}", dependencyStatus));
        });

If I try to call ActivateFetched, at **line** step, the execution never reaches that step. If I try to call ActivatedFetch at any other time, I get "ApplicationException: internal::IsInitialized() " exception

**Comment by AbhijitTaneja** [comment]
Also, this issue only happens on device, and not on unity editor

**Comment by alexames** [comment]
When you say that the "line" step never gets reached, are you saying that the call to SetDefaults never returns? Does the rest of that function run?

Do you also see this issue when running the quickstart testapp? I don't see any obvious errors in your code, so it would be good to know if it's happening in the quickstart too. If the quickstart app works, then we can start looking for things you might be doing differently from it.
